### DISCUSSION TITLE
Trigger configuration in data factory changed to allow empty files

### DISCUSSION THREAD

**Discussion by erincon01** [microsoft/finops-toolkit#528] [Ideas]
The pipeline is wisely configured to wait 60 secs before processing the .csv files.
I have a few subscriptions where data exportation first creates an empty file, and after a few seconds the file is loaded.
Having the trigger to discard empty files, makes not to tranform the .csv to .parquet.

I have changed the trigger configuration to allow empty files.

It seems like when you builded the pipeline you had in mind that takes a while to export the data, but when building the deploy, the "allow empty files" optoin was not considered.

You might find this useful for next releases.

**Comment by flanakin** [comment]
@MSBrett thoughts?

**Comment by MSBrett** [comment]
Hi @erincon01 

The delay is due to how triggers fire in ADLS.  In newer builds we no longer have a wait for the CSV files.  The trigger fires when the manifest.json file is created, waits a bit for things to quiesce and then starts processing all the files in the manifest per the concurrency settings on the for loop.

**Reply by erincon01** [reply]
@MSBrett , the trigger is not activated for json files. 
it is configured to process .csv files only.

  "blobPathBeginsWith": "/msexports/blobs/",
  "blobPathEndsWith": ".csv",

**Reply by flanakin** [reply]
@erincon01 Brett is referring to a branch where he's working on the next release. He changed it from watching for CSVs to watch for the manifest file.

**Comment by erincon01** [comment]
Hi @MSBrett , 

I've deployed v0.2, and I see you have changed the trigger name to msexports_FileAdded. 
I confirm that is subscribed to new files in the msexports folder as the older version.
It fires msexports_ExecuteETL pipeline. Older version fired: msexports_extract.

See that the trigger is disabled, and is configured to discard empty files.

I am not sure if I've deployed the correct version (deploy from here: https://microsoft.github.io/finops-toolkit/hubs).



{
    "name": "msexports_FileAdded",
    "properties": {
        "annotations": [],
        **"runtimeState": "Stopped",**
        "pipelines": [
            {
                "pipelineReference": {
                    "referenceName": "msexports_ExecuteETL",
                    "type": "PipelineReference"
                },
                "parameters": {
                    "folderName": "@triggerBody().folderPath",
                    "fileName": "@triggerBody().fileName"
                }
            }
        ],
        "type": "BlobEventsTrigger",
        "typeProperties": {
            "blobPathBeginsWith": "/msexports/blobs/",
            "blobPathEndsWith": ".csv",
            **"ignoreEmptyBlobs": true,**
            "scope": "/subscriptions/XXXXXXXXXX/resourceGroups/rg-finops-v02/providers/Microsoft.Storage/storageAccounts/finopshubpoc432XXXXXXXX",
            "events": [
                "Microsoft.Storage.BlobCreated"
            ]
        }
    }
}

**Reply by flanakin** [reply]
Note the change to monitor the manifest files didn't make it in 0.2. We're working on hopefully getting that into 0.3.

Ah, yes... There's a problem with the deploy to Azure button on that page. We use the Azure Quickstart Templates for the deployment and they have a bug where they didn't generate the correct template. I submitted a PR and am waiting on them to fix it. We're working on an update where we'll self-host the files instead of relying on them.

The way you can tell if you have the correct version is to check `config/settings.json` to see if you are running `0.2`. If you see `0.1.2`, then that's an older version. You can use this link to deploy 0.2: https://aka.ms/finops/hubs/deploy.

**Reply by BMichaelJ** [reply]
Have had a some inconsistency in deploying the Hub, first with some rights issues on CX sub, then hitting the idempotency chicken and the egg issue "RoleAssignmentUpdateNotPermitted", then with the "old" template deployment the FinopsHub page.

I have deployed 0.2 but my pipeline "msexports_ETL_ingestion" fails with error: "Operation on target Set FolderArray failed: The function 'split' expects its first parameter to be of type string. The provided value is of type 'Null'."

Am I just missing something basic? 
Any tips on what is failing?

@erincon01 You seem to have the correct pipeline changes clear, can you share recent working steps?

**Reply by erincon01** [reply]
@BMichaelJ , attached the two pipelines. it's the code i fixed for v.0.1.1.
Make sure you enable the trigger, and allow empty files.

if you cannot import the files, let me know, and I can share the json if needed.

[msexports_ETL_ingestion.zip](https://github.com/microsoft/finops-toolkit/files/14088019/msexports_ETL_ingestion.zip)
[msexports_ExecuteETL.zip](https://github.com/microsoft/finops-toolkit/files/14088020/msexports_ExecuteETL.zip)

![image](https://github.com/microsoft/finops-toolkit/assets/7208487/96b71d61-ed53-4998-83b8-48db20813a9a)

**Reply by BMichaelJ** [reply]
Thanks @erincon01 
Just to verify, this should be included in hub 0.2?
That is where my errors appear now. 
I am still working on getting a successful 0.2  greenfield installation up.

**Reply by erincon01** [reply]
Hi @BMichaelJ.

I am not sure if it should be included as I am not in the DevTeam, and I am not very familiar with github and PRs, but in my case, the changes I made in the pipeline just worked fine with my v.0.2 version.

I had to edit the Power Query transformations in regards the data filters (starDate and EndDate), but after that, all worked just fine.
If you read my comments 1-2 days ago regarding the x_*** columns, all is fixed after the FOCUS format in the exports.

**Reply by flanakin** [reply]
@BMichaelJ 0.2 had 2 bugs that have been fixed and will go out in the 0.2.1 release. There are a few other things blocking that, but you can get the release candidate from https://aka.ms/finops/hubs/rc. 

@erincon01 Thanks for sharing! These changes make hubs 0.1.1 work with 0.2 Power BI reports and trigger based on the manifest? With the 0.2 bugs resolved, we shouldn't need the former. I would point anyone who needs 0.1.1 to use the 0.1.1 Power BI reports to avoid any confusion. And Brett already has a PR out for the triggering on the manifest change. That should go out in 0.3. Is there anything else in there? 

I'm happy to setup some time to show you how to submit a PR, if you'd like 🙂 Contributions are always welcome!

**Reply by BMichaelJ** [reply]
@flanakin thanks for the rc link, we have the hub running and the reports working.

**Comment by erincon01** [comment]
I close this too. The trigger thing is fixed in the deploy you shared on friday.

**Comment by erincon01** [comment]
> I'm happy to setup some time to show you how to submit a PR, if you'd like 🙂 Contributions are always welcome!


I'd be happy to learn how to contribute. However, I am new in github :)
meanwhile, docs, or videos would help to fill the gap.

Thanks.
